cpp/src/arrow/flight/transport/ucx/ucx_server.cc

```diff
@@ -0,0 +1,646 @@
+      }
+    }
+    FLIGHT_LOG_PEER(DEBUG, peer) << "Received connection request";
+
+    auto maybe_worker = CreateWorker();
+    if (!maybe_worker.ok()) {
+      FLIGHT_LOG_PEER(WARNING, peer)
+          << "Failed to create worker" << maybe_worker.status().ToString();
+      auto status = ucp_listener_reject(listener_, request);
+      if (status != UCS_OK) {
+        FLIGHT_LOG_PEER(WARNING, peer)
+            << FromUcsStatus("ucp_listener_reject", status).ToString();
+      }
+      return;
+    }
+    auto worker = maybe_worker.MoveValueUnsafe();
+
+    // Create an endpoint to the client, using the data worker
+    {
+      ucs_status_t status;
+      ucp_ep_params_t params;
+      std::memset(&params, 0, sizeof(params));
+      params.field_mask = UCP_EP_PARAM_FIELD_CONN_REQUEST;
+      params.conn_request = request;
+
+      ucp_ep_h client_endpoint;
+
+      status = ucp_ep_create(worker->worker->get(), &params, &client_endpoint);
+      if (status != UCS_OK) {
+        FLIGHT_LOG_PEER(WARNING, peer)
+            << "Failed to create endpoint: "
+            << FromUcsStatus("ucp_ep_create", status).ToString();
+        return;
+      }
+      worker->driver.reset(new UcpCallDriver(worker->worker, client_endpoint));
+      worker->driver->set_memory_manager(memory_manager_);
+      peer = worker->driver->peer();
+    }
+
+    while (listening_.load()) {
+      auto maybe_frame = worker->driver->ReadNextFrame();
+      if (!maybe_frame.ok()) {
+        if (!maybe_frame.status().IsCancelled()) {
+          FLIGHT_LOG_PEER(WARNING, peer)
+              << "Failed to read next message: " << maybe_frame.status().ToString();
+        }
+        break;
+      }
+
+      auto status = HandleOneCall(worker->driver.get(), maybe_frame->get());
+      if (!status.ok()) {
+        FLIGHT_LOG_PEER(WARNING, peer) << "Call failed: " << status.ToString();
+        break;
+      }
+    }
+
+    // Clean up
+    auto status = worker->driver->Close();
+    if (!status.ok()) {
+      FLIGHT_LOG_PEER(WARNING, peer) << "Failed to close worker: " << status.ToString();
+    }
+    worker->worker.reset();
+    FLIGHT_LOG_PEER(DEBUG, peer) << "Disconnected";
+  }
+
+  void DriveConnections() {
+    while (listening_.load()) {
+      while (ucp_worker_progress(worker_conn_->get())) {
+      }
+      {
+        // Check for connect requests in queue
+        std::unique_lock<std::mutex> guard(pending_connections_mutex_);
+        while (!pending_connections_.empty()) {
+          ucp_conn_request_h request = pending_connections_.front();
+          pending_connections_.pop();
+
+          auto submitted = rpc_pool_->Submit([this, request]() { WorkerLoop(request); });
+          if (!submitted.ok()) {
+            ARROW_LOG(WARNING) << "Failed to submit task to handle client "
+                               << submitted.status().ToString();
+          }
+        }
+      }
+
+      if (!listening_.load()) break;
```

Comment:
I've found it's necessary in rare situations. While `Shutdown` sends `ucp_worker_signal` to "unstick" this thread from `ucp_worker_wait`, it can happen that `ucp_worker_signal` "unsticks" the thread when the thread is inside a `ucp_worker_progress` and so we can get stuck if we call `ucp_worker_wait`. I'll document this inline.